File: sdk/flutter/nhp_agent/ios/Classes/NhpAgentPlugin.swift
import Flutter
import UIKit

// Import the NhpAgent framework (compiled from Go using gomobile)
// This will be available after running scripts/build_native.sh
#if canImport(NhpAgent)
import NhpAgent
#endif

public class NhpAgentPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "io.opennhp/nhp_agent",
            binaryMessenger: registrar.messenger()
        )
        let instance = NhpAgentPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            if call.method == "close" || call.method == "stopKnockLoop" {
                // These methods don't require arguments
            } else {
                result(FlutterError(code: "INVALID_ARGS", message: "Arguments required", details: nil))
                return
            }
        }

        switch call.method {
        case "initialize":
            handleInitialize(args: call.arguments as? [String: Any] ?? [:], result: result)

        case "close":
            handleClose(result: result)

        case "setKnockUser":
            handleSetKnockUser(args: call.arguments as? [String: Any] ?? [:], result: result)

        case "addServer":
            handleAddServer(args: call.arguments as? [String: Any] ?? [:], result: result)

        case "removeServer":
            handleRemoveServer(args: call.arguments as? [String: Any] ?? [:], result: result)

        case "addResource":
            handleAddResource(args: call.arguments as? [String: Any] ?? [:], result: result)

        case "removeResource":
            handleRemoveResource(args: call.arguments as? [String: Any] ?? [:], result: result)

        case "knock":
            handleKnock(args: call.arguments as? [String: Any] ?? [:], result: result)

        case "exitResource":
            handleExitResource(args: call.arguments as? [String: Any] ?? [:], result: result)

        case "startKnockLoop":
            handleStartKnockLoop(result: result)

        case "stopKnockLoop":
            handleStopKnockLoop(result: result)

        case "generateKeys":
            handleGenerateKeys(args: call.arguments as? [String: Any] ?? [:], result: result)

        case "privateKeyToPublicKey":
            handlePrivateKeyToPublicKey(args: call.arguments as? [String: Any] ?? [:], result: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Method Handlers

    private func handleInitialize(args: [String: Any], result: @escaping FlutterResult) {
        #if canImport(NhpAgent)
        guard let workingDir = args["workingDir"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "workingDir is required", details: nil))
            return
        }
        let logLevel = args["logLevel"] as? Int ?? 2

        let success = IossdkNhpAgentInit(workingDir, logLevel)
        result(success)
        #else
        result(FlutterError(code: "NOT_COMPILED", message: "NhpAgent framework not available. Run scripts/build_native.sh first.", details: nil))
        #endif
    }

    private func handleClose(result: @escaping FlutterResult) {
        #if canImport(NhpAgent)
        IossdkNhpAgentClose()
        result(nil)
        #else
        result(nil)
        #endif
    }

    private func handleSetKnockUser(args: [String: Any], result: @escaping FlutterResult) {
        #if canImport(NhpAgent)
        let userId = args["userId"] as? String ?? ""
        let deviceId = args["deviceId"] as? String ?? ""
        let orgId = args["orgId"] as? String ?? ""
        let userData = args["userData"] as? String ?? ""

        let success = IossdkNhpAgentSetKnockUser(userId, deviceId, orgId, userData)
        result(success)
        #else
        result(FlutterError(code: "NOT_COMPILED", message: "NhpAgent framework not available", details: nil))
        #endif
    }

    private func handleAddServer(args: [String: Any], result: @escaping FlutterResult) {
        #if canImport(NhpAgent)
        guard let pubkey = args["pubkey"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "pubkey is required", details: nil))
            return
        }
        let ip = args["ip"] as? String ?? ""
        let host = args["host"] as? String ?? ""
        let port = args["port"] as? Int ?? 62206
        let expireTime = args["expireTime"] as? Int64 ?? 0

        let success = IossdkNhpAgentAddServer(pubkey, ip, host, port, expireTime)
        result(success)
        #else
        result(FlutterError(code: "NOT_COMPILED", message: "NhpAgent framework not available", details: nil))
        #endif
    }

    private func handleRemoveServer(args: [String: Any], result: @escaping FlutterResult) {
        #if canImport(NhpAgent)
        let pubkey = args["pubkey"] as? String ?? ""
        IossdkNhpAgentRemoveServer(pubkey)
        result(nil)
        #else
        result(nil)
        #endif
    }

    private func handleAddResource(args: [String: Any], result: @escaping FlutterResult) {
        #if canImport(NhpAgent)
        guard let authServiceId = args["authServiceId"] as? String,
              let resourceId = args["resourceId"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "authServiceId and resourceId are required", details: nil))
            return
        }
        let serverIp = args["serverIp"] as? String ?? ""
        let serverHost = args["serverHost"] as? String ?? ""
        let serverPort = args["serverPort"] as? Int ?? 62206

        let success = IossdkNhpAgentAddResource(authServiceId, resourceId, serverIp, serverHost, serverPort)
        result(success)
        #else
        result(FlutterError(code: "NOT_COMPILED", message: "NhpAgent framework not available", details: nil))
        #endif
    }

    private func handleRemoveResource(args: [String: Any], result: @escaping FlutterResult) {
        #if canImport(NhpAgent)
        let authServiceId = args["authServiceId"] as? String ?? ""
        let resourceId = args["resourceId"] as? String ?? ""
        IossdkNhpAgentRemoveResource(authServiceId, resourceId)
        result(nil)
        #else
        result(nil)
        #endif
    }

    private func handleKnock(args: [String: Any], result: @escaping FlutterResult) {
        #if canImport(NhpAgent)
        guard let authServiceId = args["authServiceId"] as? String,
              let resourceId = args["resourceId"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "authServiceId and resourceId are required", details: nil))
            return
        }
        let serverIp = args["serverIp"] as? String ?? ""
        let serverHost = args["serverHost"] as? String ?? ""
        let serverPort = args["serverPort"] as? Int ?? 62206

        // Perform knock on background thread to avoid blocking UI
        DispatchQueue.global(qos: .userInitiated).async {
            let jsonResult = IossdkNhpAgentKnockResource(authServiceId, resourceId, serverIp, serverHost, serverPort)
            DispatchQueue.main.async {
                result(jsonResult)
            }
        }
        #else
        result(FlutterError(code: "NOT_COMPILED", message: "NhpAgent framework not available", details: nil))
        #endif
    }

    private func handleExitResource(args: [String: Any], result: @escaping FlutterResult) {
        #if canImport(NhpAgent)
        guard let authServiceId = args["authServiceId"] as? String,
              let resourceId = args["resourceId"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "authServiceId and resourceId are required", details: nil))
            return
        }
        let serverIp = args["serverIp"] as? String ?? ""
        let serverHost = args["serverHost"] as? String ?? ""
        let serverPort = args["serverPort"] as? Int ?? 62206

        let success = IossdkNhpAgentExitResource(authServiceId, resourceId, serverIp, serverHost, serverPort)
        result(success)
        #else
        result(FlutterError(code: "NOT_COMPILED", message: "NhpAgent framework not available", details: nil))
        #endif
    }

    private func handleStartKnockLoop(result: @escaping FlutterResult) {
        #if canImport(NhpAgent)
        let count = IossdkNhpAgentKnockloopStart()
        result(count)
        #else
        result(FlutterError(code: "NOT_COMPILED", message: "NhpAgent framework not available", details: nil))
        #endif
    }

    private func handleStopKnockLoop(result: @escaping FlutterResult) {
        #if canImport(NhpAgent)
        IossdkNhpAgentKnockloopStop()
        result(nil)
        #else
        result(nil)
        #endif
    }

    private func handleGenerateKeys(args: [String: Any], result: @escaping FlutterResult) {
        #if canImport(NhpAgent)
        let cipherType = args["cipherType"] as? Int ?? 0
        let keys = IossdkNhpGenerateKeys(cipherType)
        result(keys)
        #else
        result(FlutterError(code: "NOT_COMPILED", message: "NhpAgent framework not available", details: nil))
        #endif
    }

    private func handlePrivateKeyToPublicKey(args: [String: Any], result: @escaping FlutterResult) {
        #if canImport(NhpAgent)
        let cipherType = args["cipherType"] as? Int ?? 0
        guard let privateKey = args["privateKey"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "privateKey is required", details: nil))
            return
        }
        let publicKey = IossdkNhpPrivkeyToPubkey(cipherType, privateKey)
        result(publicKey)
        #else
        result(FlutterError(code: "NOT_COMPILED", message: "NhpAgent framework not available", details: nil))
        #endif
    }
}

File: sdk/flutter/nhp_agent/android/src/main/kotlin/io/opennhp/nhp_agent/NhpAgentPlugin.kt
package io.opennhp.nhp_agent

import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*

// Import the NhpAgent library (compiled from Go using gomobile)
// This will be available after running scripts/build_native.sh
// import nhpagent.Nhpagent

/**
 * OpenNHP Agent Flutter Plugin for Android
 *
 * Provides Network-resource Hiding Protocol (NHP) functionality
 * for Zero Trust network access.
 */
class NhpAgentPlugin: FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "io.opennhp/nhp_agent")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "initialize" -> handleInitialize(call, result)
            "close" -> handleClose(result)
            "setKnockUser" -> handleSetKnockUser(call, result)
            "addServer" -> handleAddServer(call, result)
            "removeServer" -> handleRemoveServer(call, result)
            "addResource" -> handleAddResource(call, result)
            "removeResource" -> handleRemoveResource(call, result)
            "knock" -> handleKnock(call, result)
            "exitResource" -> handleExitResource(call, result)
            "startKnockLoop" -> handleStartKnockLoop(result)
            "stopKnockLoop" -> handleStopKnockLoop(result)
            "generateKeys" -> handleGenerateKeys(call, result)
            "privateKeyToPublicKey" -> handlePrivateKeyToPublicKey(call, result)
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        scope.cancel()
    }

    // MARK: - Method Handlers

    private fun handleInitialize(call: MethodCall, result: Result) {
        val workingDir = call.argument<String>("workingDir")
        val logLevel = call.argument<Int>("logLevel") ?: 2

        if (workingDir == null) {
            result.error("INVALID_ARGS", "workingDir is required", null)
            return
        }

        try {
            // TODO: Uncomment when nhpagent library is compiled
            // val success = Nhpagent.nhpAgentInit(workingDir, logLevel.toLong())
            // result.success(success)

            // Placeholder until native library is compiled
            result.error("NOT_COMPILED", "NhpAgent library not available. Run scripts/build_native.sh first.", null)
        } catch (e: Exception) {
            result.error("INIT_ERROR", e.message, null)
        }
    }

    private fun handleClose(result: Result) {
        try {
            // TODO: Uncomment when nhpagent library is compiled
            // Nhpagent.nhpAgentClose()
            result.success(null)
        } catch (e: Exception) {
            result.error("CLOSE_ERROR", e.message, null)
        }
    }

    private fun handleSetKnockUser(call: MethodCall, result: Result) {
        val userId = call.argument<String>("userId") ?: ""
        val deviceId = call.argument<String>("deviceId") ?: ""
        val orgId = call.argument<String>("orgId") ?: ""
        val userData = call.argument<String>("userData") ?: ""

        try {
            // TODO: Uncomment when nhpagent library is compiled
            // val success = Nhpagent.nhpAgentSetKnockUser(userId, deviceId, orgId, userData)
            // result.success(success)

            result.error("NOT_COMPILED", "NhpAgent library not available", null)
        } catch (e: Exception) {
            result.error("SET_USER_ERROR", e.message, null)
        }
    }

    private fun handleAddServer(call: MethodCall, result: Result) {
        val pubkey = call.argument<String>("pubkey")
        val ip = call.argument<String>("ip") ?: ""
        val host = call.argument<String>("host") ?: ""
        val port = call.argument<Int>("port") ?: 62206
        val expireTime = call.argument<Long>("expireTime") ?: 0L

        if (pubkey == null) {
            result.error("INVALID_ARGS", "pubkey is required", null)
            return
        }

        try {
            // TODO: Uncomment when nhpagent library is compiled
            // val success = Nhpagent.nhpAgentAddServer(pubkey, ip, host, port.toLong(), expireTime)
            // result.success(success)

            result.error("NOT_COMPILED", "NhpAgent library not available", null)
        } catch (e: Exception) {
            result.error("ADD_SERVER_ERROR", e.message, null)
        }
    }

    private fun handleRemoveServer(call: MethodCall, result: Result) {
        val pubkey = call.argument<String>("pubkey") ?: ""

        try {
            // TODO: Uncomment when nhpagent library is compiled
            // Nhpagent.nhpAgentRemoveServer(pubkey)
            result.success(null)
        } catch (e: Exception) {
            result.error("REMOVE_SERVER_ERROR", e.message, null)
        }
    }

    private fun handleAddResource(call: MethodCall, result: Result) {
        val authServiceId = call.argument<String>("authServiceId")
        val resourceId = call.argument<String>("resourceId")
        val serverIp = call.argument<String>("serverIp") ?: ""
        val serverHost = call.argument<String>("serverHost") ?: ""
        val serverPort = call.argument<Int>("serverPort") ?: 62206

        if (authServiceId == null || resourceId == null) {
            result.error("INVALID_ARGS", "authServiceId and resourceId are required", null)
            return
        }

        try {
            // TODO: Uncomment when nhpagent library is compiled
            // val success = Nhpagent.nhpAgentAddResource(authServiceId, resourceId, serverIp, serverHost, serverPort.toLong())
            // result.success(success)

            result.error("NOT_COMPILED", "NhpAgent library not available", null)
        } catch (e: Exception) {
            result.error("ADD_RESOURCE_ERROR", e.message, null)
        }
    }

    private fun handleRemoveResource(call: MethodCall, result: Result) {
        val authServiceId = call.argument<String>("authServiceId") ?: ""
        val resourceId = call.argument<String>("resourceId") ?: ""

        try {
            // TODO: Uncomment when nhpagent library is compiled
            // Nhpagent.nhpAgentRemoveResource(authServiceId, resourceId)
            result.success(null)
        } catch (e: Exception) {
            result.error("REMOVE_RESOURCE_ERROR", e.message, null)
        }
    }

    private fun handleKnock(call: MethodCall, result: Result) {
        val authServiceId = call.argument<String>("authServiceId")
        val resourceId = call.argument<String>("resourceId")
        val serverIp = call.argument<String>("serverIp") ?: ""
        val serverHost = call.argument<String>("serverHost") ?: ""
        val serverPort = call.argument<Int>("serverPort") ?: 62206

        if (authServiceId == null || resourceId == null) {
            result.error("INVALID_ARGS", "authServiceId and resourceId are required", null)
            return
        }

        // Perform knock on background thread
        scope.launch {
            try {
                // TODO: Uncomment when nhpagent library is compiled
                // val jsonResult = Nhpagent.nhpAgentKnockResource(
                //     authServiceId, resourceId, serverIp, serverHost, serverPort.toLong()
                // )
                // withContext(Dispatchers.Main) {
                //     result.success(jsonResult)
                // }

                withContext(Dispatchers.Main) {
                    result.error("NOT_COMPILED", "NhpAgent library not available", null)
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    result.error("KNOCK_ERROR", e.message, null)
                }
            }
        }
    }

    private fun handleExitResource(call: MethodCall, result: Result) {
        val authServiceId = call.argument<String>("authServiceId")
        val resourceId = call.argument<String>("resourceId")
        val serverIp = call.argument<String>("serverIp") ?: ""
        val serverHost = call.argument<String>("serverHost") ?: ""
        val serverPort = call.argument<Int>("serverPort") ?: 62206

        if (authServiceId == null || resourceId == null) {
            result.error("INVALID_ARGS", "authServiceId and resourceId are required", null)
            return
        }

        try {
            // TODO: Uncomment when nhpagent library is compiled
            // val success = Nhpagent.nhpAgentExitResource(
            //     authServiceId, resourceId, serverIp, serverHost, serverPort.toLong()
            // )
            // result.success(success)

            result.error("NOT_COMPILED", "NhpAgent library not available", null)
        } catch (e: Exception) {
            result.error("EXIT_RESOURCE_ERROR", e.message, null)
        }
    }

    private fun handleStartKnockLoop(result: Result) {
        try {
            // TODO: Uncomment when nhpagent library is compiled
            // val count = Nhpagent.nhpAgentKnockloopStart()
            // result.success(count.toInt())

            result.error("NOT_COMPILED", "NhpAgent library not available", null)
        } catch (e: Exception) {
            result.error("START_LOOP_ERROR", e.message, null)
        }
    }

    private fun handleStopKnockLoop(result: Result) {
        try {
            // TODO: Uncomment when nhpagent library is compiled
            // Nhpagent.nhpAgentKnockloopStop()
            result.success(null)
        } catch (e: Exception) {
            result.error("STOP_LOOP_ERROR", e.message, null)
        }
    }

    private fun handleGenerateKeys(call: MethodCall, result: Result) {
        val cipherType = call.argument<Int>("cipherType") ?: 0

        try {
            // TODO: Uncomment when nhpagent library is compiled
            // val keys = Nhpagent.nhpGenerateKeys(cipherType.toLong())
            // result.success(keys)

            result.error("NOT_COMPILED", "NhpAgent library not available", null)
        } catch (e: Exception) {
            result.error("GENERATE_KEYS_ERROR", e.message, null)
        }
    }

    private fun handlePrivateKeyToPublicKey(call: MethodCall, result: Result) {
        val cipherType = call.argument<Int>("cipherType") ?: 0
        val privateKey = call.argument<String>("privateKey")

        if (privateKey == null) {
            result.error("INVALID_ARGS", "privateKey is required", null)
            return
        }

        try {
            // TODO: Uncomment when nhpagent library is compiled
            // val publicKey = Nhpagent.nhpPrivkeyToPubkey(cipherType.toLong(), privateKey)
            // result.success(publicKey)

            result.error("NOT_COMPILED", "NhpAgent library not available", null)
        } catch (e: Exception) {
            result.error("DERIVE_KEY_ERROR", e.message, null)
        }
    }
}
